tests/test_trajectory.py

```diff
@@ -0,0 +1,305 @@
+                                        well_name = well_name,
+                                        md_datum = datum,
+                                        length_uom = 'm')
+
+    # --------- Act ----------
+    # Calculate tangent vectors based on control points
+    trajectory.set_tangents()
+
+    # --------- Assert ----------
+    assert trajectory.tangent_vectors is not None
+    assert trajectory.tangent_vectors.shape == (trajectory.knot_count, 3)
+
+
+def test_xyz_for_md(example_model_and_crs):
+
+    # --------- Arrange ----------
+    model, crs = example_model_and_crs
+    elevation = 100
+    # Create a measured depth datum
+    datum = resqpy.well.MdDatum(parent_model = model,
+                                crs_uuid = crs.uuid,
+                                location = (0, 0, -elevation),
+                                md_reference = 'kelly bushing')
+    mds = np.array([300, 310, 330, 340])
+    zs = mds - elevation
+    well_name = 'Coconut'
+    source_dataframe = pd.DataFrame({
+        'MD': mds,
+        'X': [1, 2, 3, 4],
+        'Y': [1, 2, 3, 4],
+        'Z': zs,
+        'WELL': ['Coconut', 'Coconut', 'Coconut', 'Coconut']
+    })
+
+    # Create a trajectory from dataframe
+    trajectory = resqpy.well.Trajectory(parent_model = model,
+                                        data_frame = source_dataframe,
+                                        well_name = well_name,
+                                        md_datum = datum,
+                                        length_uom = 'm')
+    # --------- Act ----------
+    # # Get the xyz triplet for a given measured depth
+    x, y, z = trajectory.xyz_for_md(md = 305)
+
+    # # -------- Assert ---------
+    assert x == y == 1.5
+    assert z == 205
+
+
+def test_splined_trajectory(example_model_and_crs):
+
+    # --------- Arrange ----------
+    model, crs = example_model_and_crs
+    elevation = 100
+    # Create a measured depth datum
+    datum = resqpy.well.MdDatum(parent_model = model,
+                                crs_uuid = crs.uuid,
+                                location = (0, 0, -elevation),
+                                md_reference = 'kelly bushing')
+    mds = np.array([300, 310, 330, 340])
+    zs = mds - elevation
+    well_name = 'Coconut'
+    source_dataframe = pd.DataFrame({
+        'MD': mds,
+        'X': [1, 2, 3, 4],
+        'Y': [1, 2, 3, 4],
+        'Z': zs,
+        'WELL': ['Coconut', 'Coconut', 'Coconut', 'Coconut']
+    })
+
+    # Create a trajectory from dataframe
+    trajectory = resqpy.well.Trajectory(parent_model = model,
+                                        data_frame = source_dataframe,
+                                        well_name = well_name,
+                                        md_datum = datum,
+                                        length_uom = 'm')
+    # --------- Act ----------
+    # Get splined trajectory
+    splined_trajectory = trajectory.splined_trajectory(well_name = 'Coconut')
```

Comment:
As your basic trajectory data is slightly impossible, I'm not totally sure of the behaviour of the splined_trajectory() method in these conditions.

Reply:
My thought process was that the splined trajectory end points should be almost equal to the original trajectory end points so that's why I added those assertions. Was that assumption completely off-base?